docs/transformations.md

```diff
@@ -0,0 +1,280 @@
+    activations = jnp.tanh(outputs)
+  return outputs
+```
+
+We often instead write `jnp.dot(inputs, W)` to allow for a batch dimension on the
+left side of `inputs`, but we’ve written this particular prediction function to
+apply only to single input vectors. If we wanted to apply this function to a
+batch of inputs at once, semantically we could just write
+
+```{code-cell}
+:tags: [hide-cell]
+# Create some sample inputs & parameters
+import numpy as np
+k, N = 10, 5
+input_batch = np.random.rand(k, N)
+params = [
+  (np.random.rand(N, N), np.random.rand(N)),
+  (np.random.rand(N, N), np.random.rand(N)),
+]
+```
+
+```{code-cell}
+from functools import partial
+predictions = jnp.stack(list(map(partial(predict, params), input_batch)))
+```
+
+But pushing one example through the network at a time would be slow! It’s better
+to vectorize the computation, so that at every layer we’re doing matrix-matrix
+multiplication rather than matrix-vector multiplication.
+
+The {func}`vmap` function does that transformation for us. That is, if we write:
+
+```{code-cell}
+from jax import vmap
+predictions = vmap(partial(predict, params))(input_batch)
+# or, alternatively
+predictions = vmap(predict, in_axes=(None, 0))(params, input_batch)
+```
+
+then the {func}`vmap` function will push the outer loop inside the function, and our
+machine will end up executing matrix-matrix multiplications exactly as if we’d
+done the batching by hand.
+
+It’s easy enough to manually batch a simple neural network without {func}`vmap`, but
+in other cases manual vectorization can be impractical or impossible. Take the
+problem of efficiently computing per-example gradients: that is, for a fixed set
+of parameters, we want to compute the gradient of our loss function evaluated
+separately at each example in a batch. With {func}`vmap`, it’s easy:
+
+```{code-cell}
+:tags: [hide-cell]
+# create a sample loss function & inputs
+def loss(params, x, y0):
+  y = predict(params, x)
+  return jnp.sum((y - y0) ** 2)
+
+inputs = np.random.rand(k, N)
+targets = np.random.rand(k, N)
+```
+
+```{code-cell}
+per_example_gradients = vmap(partial(grad(loss), params))(inputs, targets)
+```
+
+Of course, {func}`vmap` can be arbitrarily composed with {func}`jit`, {func}`grad`,
+and any other JAX transformation! We use {func}`vmap` with both forward- and reverse-mode automatic
+differentiation for fast Jacobian and Hessian matrix calculations in
+{func}`jax.jacfwd`, {func}`jax.jacrev`, and {func}`jax.hessian`.
+
+## SPMD programming with `pmap`
+
+For parallel programming of multiple accelerators, like multiple GPUs, use
+{func}`jax.pmap`.
+With {func}`pmap` you write single-program multiple-data (SPMD) programs, including
+fast parallel collective communication operations. Applying {func}`pmap` will mean
+that the function you write is compiled by XLA (similarly to {func}`jit`), then
+replicated and executed in parallel across devices.
+
+Here's an example on an 8-GPU machine:
+
+```python
```

Comment:
Note `python` here rather than `{code-cell}` which means that this code does not get executed or tested when the docs are built. The RTD build is on a single-core machine, so pmap fails.

Is there any way we could emulate a multi-core machine to test pmap in that environment?

Reply:
Managed to do it using `--xla_force_host_platform_device_count=8` in a hidden setup cell.